lightningd/gossip/gossip.c

```diff
@@ -257,72 +257,57 @@ static bool has_even_bit(const u8 *bitmap)
 	return false;
 }
 
-static struct io_plan *recv_client_req(struct io_conn *conn, struct peer *peer);
-static struct io_plan *client_req_in(struct io_conn *conn, struct peer *peer)
+/**
+ * owner_msg_in - Called by the `peer->owner_conn` upon receiving a
+ * message
+ */
+static struct io_plan *owner_msg_in(struct io_conn *conn,
+				    struct daemon_conn *dc)
 {
-	/* TODO(cdecker) Handle incoming requests */
-	status_trace("Received message from client %s", tal_hexstr(peer, peer->proxy_in, tal_count(peer->proxy_in)));
-	return recv_client_req(conn, peer);
-}
+	struct peer *peer = container_of(dc, struct peer, owner_conn);
+	u8 *msg = dc->msg_in;
 
-static struct io_plan *client_pkt_out(struct io_conn *conn, struct peer *peer);
-static struct io_plan *recv_client_req(struct io_conn *conn, struct peer *peer)
-{
-	return io_read_wire(conn, peer, &peer->proxy_in, client_req_in, peer);
+	int type = fromwire_peektype(msg);
+	if (type == WIRE_CHANNEL_ANNOUNCEMENT || type == WIRE_CHANNEL_UPDATE ||
+	    type == WIRE_NODE_ANNOUNCEMENT) {
+		handle_gossip_msg(peer->daemon->rstate, dc->msg_in);
+	}
+	return daemon_conn_read_next(conn, dc);
 }
 
-static struct io_plan *client_dump_gossip(struct io_conn *conn, struct peer *peer)
+/**
+ * nonlocal_dump_gossip - catch the nonlocal peer up with the latest gossip.
+ *
+ * Registered as `msg_queue_cleared_cb` by the `peer->owner_conn`.
+ */
+static struct io_plan *nonlocal_dump_gossip(struct io_conn *conn, struct daemon_conn *dc)
 {
 	struct queued_message *next;
+	struct peer *peer = container_of(dc, struct peer, owner_conn);
+
+
+	/* Make sure we are not connected directly */
+	if (peer->local)
```

Comment:
Not sure, nothing prevents us from enqueuing messages destined for a future owner, and if it is small enough we get a queue cleared, at which point `nonlocal_dump_gossip` gets called. So making this non-fatal seems to be better.

in `pkt_out` on the other hand we should not be enqueuing new messages once we handed off the peer, so I added an `assert` there.